### PR TITLE
#889 Fix flaky TestCraneWithoutDockerServer.normalCase

### PR DESCRIPTION
**Pull request by chia7712**
### Checklist
- [x] [Review Contribution Guidelines](https://github.com/oharastream/ohara/blob/master/CONTRIBUTING.md)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#authors))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #889)
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
the loop tests are [here](https://builds.is-land.com.tw/job/PreTest-OHARA/342/console)
- [x] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

**Comment by chia7712** [comment]
@saivirtue Please take a look at this RP

File: ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala [hunk 1]
```diff
@@ -74,12 +74,13 @@ class TestCraneWithoutDockerServer extends SmallTest with Matchers {
...
-
-    info._1.name shouldBe WAREHOUSE_NAME
-    info._2.size shouldBe SIZE
+    CommonUtils.await(
+      () => {
+        val info = awaitResult(crane.get(WAREHOUSE_NAME))
+        info._1.name == WAREHOUSE_NAME && info._2.size == SIZE
+      },
+      java.time.Duration.ofSeconds(20)
+    )
```
Comment (saivirtue) [inline ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala]: Does this mean thread will wait till `info` get all the result?

Comment (chia7712) [reply ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala]: yep. If our cache is powerful to update the result quickly, the end of wait ensues. By contrast, we give 20 seconds to wait the **slow** cache. It is still time-based test but is more stable.

Comment (saivirtue) [reply ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala]: so, here the result of `a == b` is equals to `a shouldBe b` as previous?

Comment (chia7712) [reply ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala]: Not really. "a == b" just return a boolean value but "a shouldBe b" produce a exception if comparison result is false.

Comment (saivirtue) [reply ohara-agent/src/test/scala/com/island/ohara/agent/docker/TestCraneWithoutDockerServer.scala]: got it

File: ohara-agent/src/main/scala/com/island/ohara/agent/fake/FakeDockerClient.scala [hunk 2]
```diff
@@ -16,26 +16,32 @@
...
 
-import scala.collection.mutable
 import scala.concurrent.{ExecutionContext, Future}
-
+import scala.collection.JavaConverters._
 private[agent] class FakeDockerClient extends ReleaseOnce with DockerClient {
   private val LOG = Logger(classOf[FakeDockerClient])
   private[this] val FAKE_KIND_NAME: String = "FAKE"
-  private[this] val cachedContainers = new mutable.HashMap[String, ContainerInfo]()
+  private[this] val cachedContainers = new ConcurrentHashMap[String, ContainerInfo]()
```
Comment (saivirtue) [inline ohara-agent/src/main/scala/com/island/ohara/agent/fake/FakeDockerClient.scala]: overall +1. 
Another small question, if we use `ConcurrentHashMap`, do we still need using `lock.synchronized` ?

Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/fake/FakeDockerClient.scala]: > if we use ConcurrentHashMap, do we still need using lock.synchronized ?

If the method operates only the ConcurrentHashMap, we don't need to sync on method. However, the lock is still required when we do multiple operations in the method. For example,
```scala
   val a = ConcurrentHashMap.get(x)
   val b = process(a)
   ConcurrentHashMap.put(b)
``` 

In order to keep data consistency, we ought to add sync block for above case.

```scala
  sync {
    val a = ConcurrentHashMap.get(x)
    val b = process(a)
    ConcurrentHashMap.put(b)
  }
```

Comment (chia7712) [reply ohara-agent/src/main/scala/com/island/ohara/agent/fake/FakeDockerClient.scala]: At any rate, the code are in testing scope. It is not necessary to make a perfect fake stuff :)

**Pull request merged**